packages/bridge-ui/src/erc20/bridge.ts

```diff
@@ -0,0 +1,104 @@
+import { BigNumber, Contract, Signer } from "ethers";
+import type { Transaction } from "ethers";
+import type { ApproveOpts, Bridge, BridgeOpts } from "../domain/bridge";
+import TokenVault from "../constants/abi/TokenVault";
+import ERC20 from "../constants/abi/ERC20";
+
+class ERC20Bridge implements Bridge {
+  private async spenderRequiresAllowance(
+    tokenAddress: string,
+    signer: Signer,
+    amount: BigNumber,
+    bridgeAddress: string
+  ): Promise<boolean> {
+    const contract: Contract = new Contract(tokenAddress, ERC20, signer);
+    const allowance: BigNumber = await contract.allowance(
+      await signer.getAddress(),
+      bridgeAddress
+    );
+
+    return allowance.lt(amount);
+  }
+
+  async RequiresAllowance(opts: ApproveOpts): Promise<boolean> {
+    return await this.spenderRequiresAllowance(
+      opts.contractAddress,
+      opts.signer,
+      opts.amountInWei,
+      opts.spenderAddress
+    );
+  }
+
+  async Approve(opts: ApproveOpts): Promise<Transaction> {
+    if (
```

Comment:
I am a little confused about this `if`. Do we need this here? From what I understand we are calling `Approve` when we the `spender` doesn't have enough allowance so wouldn't this prevent the user from approving?

Reply:
yes you are correct, this needs a "!" at the start!